### PR TITLE
Describe testing in the hacking doc

### PR DESCRIPTION
**Pull request by rgrinberg**
Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

### CODE REVIEW

File: HACKING.md [hunk 1]
```diff
@@ -54,6 +54,44 @@ quite pleasant.
...
+A description of the test. The command running the tests is preceeded by
+two spaces and a $. The expected output of the command is also indetended by
+two spaces and is right below the command (note the absence of a $)
+
+  $ echo "the expected output is below"
+  the expected output is below
+```
+
+Running the entire test suite is done with `$ make test`. A particular test can
+be executed with `$ ./boot.exe build @<test-name>`.
```
Comment (diml) [inline HACKING.md]: boot.exe doesn't support that. I wonder if we shouldn't just recommend to call `dune build ...` rather that give custom/make commands. It seems better to encourage to use dune directly.

Comment (rgrinberg) [reply HACKING.md]: Ah yeah, indeed. I actually just use an alias for the dev version of dune myself. Does it make sense to just add a `$ make dune <args>` target that will delegate to dune?

**Comment by rgrinberg** [comment]
I added the `$ make dune ..` target to execute the locally built dune.

File: HACKING.md (hunk 1 above)
Comment (diml) [reply HACKING.md]: I was thinking of just using an installed dune binary

Comment (rgrinberg) [reply HACKING.md]: That also has some disadvantages. Like having to make sure the dune binary is installed. But ok, I'll change it to that. I'll keep the new make target if you don't mind.

File: HACKING.md [hunk 2]
```diff
@@ -54,6 +54,44 @@ quite pleasant.
...
+tests for its test suite. The test suite is contained in
+[test/blackbox-tests](test/blackbox-tests). A single test consists of a test
+directory in the test-cases/ sub directory which contains a run.t file defining
+the test.
+
+An example `run.t` file:
+
+```
+A description of the test. The command running the tests is preceeded by
+two spaces and a $. The expected output of the command is also indetended by
```
Comment (emillon) [inline HACKING.md]: typo: indented

File: HACKING.md [hunk 3]
```diff
@@ -54,6 +54,44 @@ quite pleasant.
...
+be executed with `$ dune build @<test-name>`.
+
+Running the test will execute the command after the `$` and its output will be
+compared against the expected output right below the command. Any differences
+will result in a test failure.
+
+### Adding a Test
+
+Simply add a new directory in test/blackbox-tests/test-cases and then `$ make`
+generate the rules for the test , followed by `$ make accept-corrections` to
```
Comment (emillon) [inline HACKING.md]: I personally use `make promote` which is an alias. Is it worth simplifying and have a single target?

Comment (diml) [reply HACKING.md]: Yeah, that seems worth it. Otherwise the reader might wonder what the difference between the two is

**Comment by rgrinberg** [comment]
Thanks for the careful review. Suggestions on how to expand the document are welcome as well.

**Pull request merged**